src/Microsoft.AspNetCore.Authentication.MicrosoftAccount/MicrosoftAccountOptions.cs

```diff
@@ -30,9 +30,7 @@ public MicrosoftAccountOptions()
             ClaimResolvers.Add(ClaimTypes.Name, "displayName");
             ClaimResolvers.Add(ClaimTypes.GivenName, "givenName");
             ClaimResolvers.Add(ClaimTypes.Surname, "surname");
-            // Last one wins, if present
-            ClaimResolvers.Add(ClaimTypes.Email, "userPrincipalName");
-            ClaimResolvers.Add(ClaimTypes.Email, "mail");
+            ClaimResolvers.AddCustom(ClaimTypes.Email, user => user.Value<string>("mail") ?? user.Value<string>("userPrincipalName"));
```

Comment:
Well for code like this I'd prefer if it was self describing, i.e. if it was JsonPayloadClaimsMap.Add("ClaimsTypes.Email", "userPrincipalName) that has much more context